File: 12-1.rb
#!/usr/bin/env ruby

instructions = File.read('12.input').lines.map(&:strip)

instruction_format = /(?<inst>N|S|E|W|L|R|F)(?<oper>[[:digit:]]+)/

class Ship
  attr_accessor :direction
  attr_accessor :x
  attr_accessor :y

  def initialize(x, y)
    @x = x
    @y = y
    @direction = 90
  end

  def forward(steps)
    case @direction
    when 0
      move(0, steps)
    when 90
      move(steps, 0)
    when 180
      move(0, -steps)
    when 270
      move(-steps, 0)
    else
      print "Illegal direction: #{@direction}\n"
      exit
    end
  end

  def move(x, y)
    @x += x
    @y += y
  end

  def turn(direction, angle)
    angle = -angle if direction == 'L'
    @direction = (@direction + angle) % 360
  end

  def to_s
    "<#{self.class}: (#{@x}, #{@y}), #{@direction}>"
  end
end

ship = Ship.new(0, 0)

instructions.each do |insn|
  instruction_format.match(insn) do |m|
    case m['inst']
    when 'N'
      ship.move(0, m['oper'].to_i)
    when 'S'
      ship.move(0, -m['oper'].to_i)
    when 'E'
      ship.move(m['oper'].to_i, 0)
    when 'W'
      ship.move(-m['oper'].to_i, 0)
    when 'L'
      ship.turn(m['inst'], m['oper'].to_i)
    when 'R'
      ship.turn(m['inst'], m['oper'].to_i)
    when 'F'
      ship.forward(m['oper'].to_i)
    else
      print "Illegal instruction: #{insn}\n"
      exit
    end
  end
end

print "#{ship.x.abs + ship.y.abs}\n"

File: 12-1.py
#!/usr/bin/env python

import re
import sys

instructions = [insn.strip() for insn in open('12.input').readlines()]

instruction_format = re.compile("(?P<inst>N|S|E|W|L|R|F)(?P<oper>[0-9]+)")

class Ship:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.direction = 90

    def forward(self, steps):
        if self.direction == 0:
            self.move(0, steps)
        elif self.direction == 90:
            self.move(steps, 0)
        elif self.direction == 180:
            self.move(0, -steps)
        elif self.direction == 270:
            self.move(-steps, 0)
        else:
            print(f"Illegal direction: {self.direction}")
            sys.exit()

    def move(self, x, y):
        self.x += x
        self.y += y

    def turn(self, direction, angle):
        if direction == 'L':
            angle = -angle
        self.direction = (self.direction + angle) % 360

    def __str__(self):
        return f"<{self.__class__.__name__}: ({self.x}, {self.y}), {self.direction}>"

ship = Ship(0, 0)

for insn in instructions:
    m = instruction_format.match(insn)
    if m['inst'] == 'N':
        ship.move(0, int(m['oper']))
    elif m['inst'] == 'S':
        ship.move(0, -int(m['oper']))
    elif m['inst'] == 'E':
        ship.move(int(m['oper']), 0)
    elif m['inst'] == 'W':
        ship.move(-int(m['oper']), 0)
    elif m['inst'] == 'L':
        ship.turn(m['inst'], int(m['oper']))
    elif m['inst'] == 'R':
        ship.turn(m['inst'], int(m['oper']))
    elif m['inst'] == 'F':
        ship.forward(int(m['oper']))
    else:
      print(f"Illegal instruction: {insn}")
      sys.exit()

print(f"{abs(ship.x) + abs(ship.y)}")
